### PR TITLE
hsa profiler

### PR DESCRIPTION
**Pull request by nimlgen**

### CODE REVIEW

**Comment by nimlgen** [comment]
@geohot, looks good overall?

**Comment by geohot** [comment]
Yes, want to post a screenshot of it working in thread? I'll test now

**Comment by geohot** [comment]
Screenshot 2024-03-13 at 8 38 46 AM

**Comment by geohot** [comment]
Hmm, actually there's a lot of overhead caused by the profiler. Anything we can do about this?
Screenshot 2024-03-13 at 8 43 39 AM

**Comment by geohot** [comment]
Screenshot 2024-03-13 at 8 56 21 AM

Overhead on LLaMA. Ideally this is sub 5%

File: tinygrad/runtime/graph/hsa.py [hunk 1]
```diff
@@ -145,6 +151,7 @@ def __call__(self, input_rawbuffers: List[Buffer], var_vals: Dict[Variable, int]
       hsa.hsa_signal_wait_scacquire(self.finish_signal, hsa.HSA_SIGNAL_CONDITION_LT, 1, (1 << 64) - 1, hsa.HSA_WAIT_STATE_ACTIVE)
       et = time.perf_counter() - st
 
+    for prof in self.profile_info: Profiler.track(*prof)
```
Comment (geohot) [inline tinygrad/runtime/graph/hsa.py:154]: This looks slow, maybe that's the python overhead? Why isn't it going right to self.tracked_signals?

**Comment by geohot** [comment]
CIFAR looks better, but llama is still high overhead. What's the difference?

**Comment by nimlgen** [comment]
Ugh structs... llama has more kernels to profile, so it's a bit worse than cifar

```
19   67.45 ms run,   16.52 ms python,   50.93 ms HSA * 6, 1072.75 loss, 0.000086 LR, 12.86 GB used,  29867.48 GFLOPS,   2014.44 GOPS
```
```
enqueue in   8.09 ms
total  39.17 ms, 25.53 tok/sec
```

**Comment by geohot** [comment]
Yea, struct creation is stupidly slow. Are we doing any struct creation in HIP?

**Comment by geohot** [comment]
Merged, I think usable as is. @uuuvn maybe this can help answer questions why ring reduce is slower?

**Pull request merged**